Feat/Data Pipeline Rework

yeshavlamor opened the pull request:
## Summary
Overhaul scheme processing pipeline: deprecate dataset_workflow in favor of data pipeline hosted in GCP's Cloud Run + Firestore Vector Search architecture.

Apologies for the big ass PR. 

## Problems Solved

### 1. Vector store index Management
Our ChromaDB setup stores indexes in memory, which means managing reindexing pipelines with updated indexes was tricky. We can inject it using github actions but that would means reindexing can only happen if there is a redeployment of backend. 2nd option is we store chromaDB index in GCS bucket and pull when we need it, but it will cause cold starts. **Solution**: Migrate to Firestore Vector Search - API-based, no index management needed. I expect minimal impact on accuracy because specs are pretty much the same just that the embeddings dimensions is 2k instead of 3k, still using cosine similarity.  

### 2. Human-in-the-Loop for Scheme Quality
No reliable process for reviewing scheme quality and link validity before. **Solution**: Slack approval workflow for new schemes + monthly scheduled link checker with Slack notifications. Kudos to @yeshavlamor for starting this.

### 3. Incomplete Field Extraction
Previous basic Selenium crawler couldn't deep crawl, missing fields like address, images, and contact numbers often found on subpages. **Solution**: crawl4ai with BFS deep crawl strategy + pydoll for Cloudflare bypass.

## Architecture Changes

### New Scheme Submission Flow
1. User submits new scheme via `/contribute` page (simplified to only accept new schemes)
2. Frontend creates document in `schemeEntries` collection → returns immediately
3. Firestore trigger calls `scheme-processor` Cloud Run service (async, non-blocking)
4. Cloud Run scrapes website, extracts fields via LLM, posts to Slack
5. Admin reviews scheme in Slack modal (can edit fields before approval)
6. On approval, scheme added to `schemes` collection (main collection)

### Scheduled Batch Job (Monthly)
- Checks all scheme links for validity
- Marks dead links as inactive, posts summary to Slack
- Reindexes embeddings into Firestore composite index

### Deprecated
- `dataset_workflow/` - No longer needed for scheme ingestion
- ChromaDB and related ML dependencies (torch, transformers, spacy, faiss)

## Key Components

### scheme-processor (Cloud Run)
FastAPI service (already deployed in dev acc)

**Dual Crawler Strategy:**
- **crawl4ai**: Primary crawler using BFS deep crawl strategy:
  - `max_depth=1`: Crawls links 1 level deep from main page
  - `max_pages=5`: Limits to 5 pages per scheme
  - `include_external=False`: Only follows internal links
  - Aggregates content from multiple pages (e.g., eligibility, FAQ, contact) before LLM extraction
- **pydoll**: Fallback for Cloudflare-protected sites using headless Chrome with virtual display (Xvfb) to bypass bot detection.

**LLM Extraction:** Structured JSON output with Pydantic schema for reliable field extractions

**Security:** Firebase Functions authenticate to Cloud Run via service account identity tokens (IAM `roles/run.invoker`).

### Other Components
- `functions/batch_jobs/` - Scheduled link check and reindex
- `functions/new_scheme/` - Firestore trigger and Slack approval handlers

## Ad-hoc Scripts
Located in `functions/scripts/`, useful for manual operations outside scheduled jobs:
- `run_link_check_and_reindex.py` - Run link validation and reindex manually
- `populate_embeddings.py` - Bulk generate embeddings for schemes
- `test_vector_search.py` - Test vector search queries

```bash
cd backend/functions
uv run python -m scripts.<script_name>
```

## Test Plan
- [x] Submit new scheme via frontend (returns immediately)
- [x] Verify Slack notification with Review/Reject buttons
- [x] Approve scheme and verify it appears in schemes collection
- [x] Verify scheme search works with Firestore vector search
- [x] Run link check batch job manually

## Screenshots
<img width="579" height="571" alt="image" src="https://github.com/user-attachments/assets/62c0fa00-0c52-4fc7-afeb-142d92196390" />
<img width="675" height="670" alt="image" src="https://github.com/user-attachments/assets/d0372842-3999-47ca-a424-d07c92387b4a" />
<img width="515" height="631" alt="image" src="https://github.com/user-attachments/assets/941bfaad-7007-4acc-8b3f-777be9fe3e4e" />
<img width="659" height="766" alt="image" src="https://github.com/user-attachments/assets/6897f6d8-ce78-4e78-937b-557722756baa" />

yevkim committed:
initial commit

yevkim committed:
Update readme and scripts for export from dev and import to prod

yevkim committed:
add a export function to get prod user logs as csv

Commit:
add new fields to be extracted

IsaacChanXi committed:
add script for adding fields to fire store

IsaacChanXi committed:
added script for creating indexes and models

yevkim committed:
processing expanded dataset

IsaacChanXi committed:
erge branch 'dataset-workflow' of https://github.com/bettersg/SchemesSG_v2 into add-auto-extraction-to-scraping

IsaacChanXi committed:
add shell script and code for uploading artefacts to storage

IsaacChanXi committed:
add python run to shell script

yevkim committed:
Updated the key schemes file

IsaacChanXi committed:
add code based on latest feedback

yevkim committed:
Merge pull request #167 from bettersg/add-auto-extraction-to-scraping

Add auto extraction to scraping

yevkim committed:
Clean up, improve robustness error loggin, add pdf handling scraping, test overall sh script on dev with 400 schemes

yevkim committed:
leverage system update time instead of custom field, fix the uplooad file for zip, helper function to remove field in documents

IsaacChanXi committed:
improve extraction pipeline with examples and chunking

IsaacChanXi committed:
add error handling for fields with no scraped text

IsaacChanXi committed:
add doc-strings for methods

yevkim committed:
Merge pull request #171 from bettersg/add-auto-extraction-to-scraping

Add auto extraction to scraping

yevkim committed:
update the export import functions for dataset workflow

yevkim committed:
Add new step to extract summary and planning area and populate in schemes

yevkim committed:
Remove the token from commit and move to .env

yevkim committed:
Improve scraping, support multiple physical locations per schemes, add service area

yevkim committed:
additional logic for scraping

yevkim committed:
rename add town area to focus on town area only

yevkim committed:
updates for new scheme add for carecorner

IsaacChanXi committed:
add python pipeline for execution

IsaacChanXi committed:
refactor code for easier imports

IsaacChanXi committed:
add refactoring of workflow into docker image

IsaacChanXi committed:
uncomment pipeline code

IsaacChanXi committed:
remove unused csv files and run steps shell script

IsaacChanXi committed:
remove error logging to csv files

IsaacChanXi committed:
remove unused csv files

IsaacChanXi committed:
add back log error to csv for Main scrape py

IsaacChanXi committed:
add passing of doc ids to process via cmd line

IsaacChanXi committed:
add last time an id was scraped successfully

IsaacChanXi committed:
add refresh condition if update is more than a month ago

yevkim committed:
Merge pull request #221 from bettersg/refactor_shell_script_data_pipeline

Refactor shell script data pipeline

yevkim committed:
Update the logic to run locally and reduce further errors

yevkim committed:
adding centralized logging to have one audit file recording all terminal output

yevkim committed:
Improve scraping, logging, and verifying completeness

yevkim committed:
Add more checks for count documents

yevkim committed:
add cleaning of llm description to avoid escape characters

wtlow003 committed:
feat(dataset_workflow): initial changes for chroma db artefacts generation

wtlow003 committed:
fix(dataset-workflow): zip file path

wtlow003 committed:
feat: add test chroma artefacts pipeline step

wtlow003 committed:
fix: use upsert semantics instead of add to update existing ids in future

wtlow003 committed:
fix: explicit conversion from cosine distances to similarities

yevkim committed:
Merge pull request #233 from bettersg/stg

Release October 2025

kairayzo committed:
Merge pull request #240 from bettersg/stg

Added about section

Link to Issue:

#234

Changes

Removed /about route in header
Added revamped about section in landing page below query bar
Updated SchemesSG logo and added SchemesSG favicon
Removed footer section
Fixed Geist font not applying

kairayzo committed:
Removed nonexistent contact email from about section

yevkim committed:
Merge pull request #241 from bettersg/feat/234-about-section

Removed nonexistent contact email from about section

kairayzo committed:
Added Care Corner to partners.tsx

yevkim committed:
Merge pull request #243 from bettersg/feat/234-about-section

Added Care Corner to partners.tsx

yeshavlamor committed:
Merge remote-tracking branch 'origin/refactor/chroma-db-setup'

yeshavlamor committed:
initial commit

yeshavlamor committed:
slack integration using firebase functions

longwind48 reviewed (changes requested):
Great job @yeshavlamor , but I need you to:
- clean up irrelevant directories and files, e.g. dataset_workflow/
- supplement your description so that I know how to run your setup locally in my laptop. Include all the links and add me to the slack channel so i can see your tests. Pass me the updated .env too, if any.

yeshavlamor committed:
add readme file for setup instructions

yeshavlamor committed:
delete unrelated directories

longwind48 committed:
Add new scheme submission pipeline with crawl4ai

- Add pipeline_runner.py with crawl4ai for web crawling and LLM extraction
- Add smart logo selection prioritizing header/footer images
- Add duplicate domain validation to prevent duplicate submissions
- Add Slack review modal with image preview and approval workflow
- Add approval handler to create schemes and update schemeEntries

longwind48 committed:
Add Slack handlers for new scheme review workflow

- Add review_new_scheme button handler to open approval modal
- Add reject_new_scheme button handler for quick rejection
- Add preview_image_button handler for dynamic image preview
- Add new_scheme_approval_submit handler for modal submissions
- Improve error logging with detailed Slack API error messages

longwind48 committed:
Add dependencies and config for new scheme pipeline

- Add crawl4ai, playwright dependencies to requirements.txt
- Add Playwright browser dependencies to Dockerfile
- Export new scheme trigger function in main.py
- Update firebase.json with function configuration

longwind48 committed:
Update Docker and logging configuration

- Add environment variables to docker-compose for Azure OpenAI
- Update start.sh for emulator configuration
- Improve logging setup with loguru configuration

longwind48 committed:
Add Firestore data exports to gitignore

- Ignore prod_schemes_data.json and dev_schemes_data.json
- Ignore firestore-backup directory

longwind48 committed:
Simplify contribute page to only accept new schemes

longwind48 committed:
Change duplicate check from domain to full URL path

longwind48 committed:
Add google_sheets_credentials.json to gitignore

longwind48 committed:
Add scheme-processor Cloud Run service with identity token auth

longwind48 committed:
Add deploy script with IAM setup for Cloud Run

longwind48 committed:
Add scheme-processor service to docker-compose for local dev

longwind48 committed:
Move constants to dedicated module in new_scheme

longwind48 committed:
Improve new scheme submission flow with async processing and image preview error handling

longwind48 committed:
Remove dead code - processing moved to scheme-processor Cloud Run

longwind48 committed:
feat: add scheduled link check and reindex batch job

- Add monthly scheduled job for link validation
- Check all scheme links and mark dead ones inactive
- Post summary to Slack with dead link details
- Trigger embedding reindex after link check

longwind48 committed:
chore: update dependencies for Firestore Vector Search

- Remove chromadb, faiss-cpu, spacy, torch, transformers
- Add google-cloud-firestore for native vector search
- Add litellm for LLM abstraction
- Add pydoll-python for Cloudflare bypass

longwind48 committed:
feat: migrate search to Firestore native vector search

- Update searchModelManager for Firestore vector search
- Add NaN/Infinity handling in JSON encoder
- Remove ChromaDB dependency

longwind48 committed:
chore: update build and config files

- Update Dockerfile.firebase for new dependencies
- Add firebase.json configuration
- Update start.sh for emulator settings
- Add dependencies to pyproject.toml

longwind48 committed:
feat: add link check and reindex utility modules

- Add check_link.py for validating scheme URLs
- Add reindex_embeddings.py for vector search reindexing

longwind48 committed:
chore: remove outdated ChromaDB test file

longwind48 committed:
chore: add Firestore indexes configuration

longwind48 committed:
chore: add CLAUDE.md to gitignore

longwind48 committed:
feat: add ad-hoc scripts for vector search operations

- Add populate_embeddings.py for bulk embedding generation
- Add run_link_check_and_reindex.py for manual batch job runs
- Add test_vector_search.py for testing vector search queries

longwind48 committed:
docs(README): update backend documentation with architecture and setup details

longwind48 committed:
docs: Rewrite backend README with new architecture

- Add architecture overview with Cloud Run + Firebase Functions
- Document new scheme submission flow with Slack approval
- Add directory structure with all modules explained
- Include Docker Compose setup instructions
- Document API endpoints, batch jobs, and ad-hoc scripts
- Add scheme-processor Cloud Run service documentation
- Fix ReDoS vulnerability in email regex extraction

longwind48 committed:
Add ruff config and fix linting issues
